src/identities/mnemonic.ts

```diff
@@ -152,21 +174,25 @@ export default class Mnemonic extends Identity implements IdentityInterface {
         confidentialAddress: confidentialAddress!,
         blindingPrivateKey: blindingKeyPair.privateKey!.toString('hex'),
       },
-      derivationPath: `${this.baseDerivationPath}/${currentIndex}`,
+      derivationPath: `${this.baseDerivationPath}/${index}`,
       signingPrivateKey: signingKeyPair.privateKey!.toString('hex'),
     };
-    // store the generation inside local cache
-    this.scriptToAddressCache.set(script, newAddressGeneration);
     // return the generation data
-    return newAddressGeneration.address;
+    return newAddressGeneration;
   }
 
   getNextAddress(): AddressInterface {
-    return this.getAddress(false);
+    const addr = this.getAddress(false);
+    this.persistAddressToCache(addr);
+    this.changeIndex = this.index;
```

Comment:
This will increment the changeIndex, even if not needed. There can be an occasione where we derive like many external addresses, but we re-use the same change address.  Only the getNextChangeAddress should increment that, not other functions.